vnext/Microsoft.ReactNative.SharedManaged/JSValueReaderGenerator.cs

```diff
@@ -12,12 +13,30 @@
 
 namespace Microsoft.ReactNative.Managed
 {
-  static class JSValueReaderGenerator
+  public static class JSValueReaderGenerator
   {
+    private static readonly ConcurrentBag<Assembly> m_registerdAssemblies = new ConcurrentBag<Assembly>();
     private static readonly Lazy<KeyValuePair<Type, MethodInfo>[]> s_allMethods;
     private static readonly Lazy<IReadOnlyDictionary<Type, MethodInfo>> s_nonGenericMethods;
     private static readonly Lazy<IReadOnlyDictionary<Type, SortedList<Type, MethodInfo>>> s_genericMethods;
 
+    public static void RegisterAssembly(Assembly assembly)
+    {
+      // If already registered, no need to add
+      if (m_registerdAssemblies.Contains(assembly))
+      {
+        return;
+      }
+
+      // If we are going to add the assembly, we can't after we have started reading and processing the methods.
+      if (s_allMethods.IsValueCreated)
+      {
+        throw new InvalidOperationException("Cannot register assemblies dynamically after the first value is serialized.");
+      }
+
+      m_registerdAssemblies.Add(assembly);
```

Comment:
What happens if some other thread won the race and already added it?
I am generally skeptical about the concurrent data structures because like in this case it really feels that we need a lock for the whole body of this function to ensure thread safety. E.g. checking IsValueCreated has no meaning before we call Add. The value can be created right between the check and the call to Add. It is not safe.

Reply:
The worst that could happen is that an assembly ends up multiple times in the bag which will have a slight perf issues for enumerating members (I didn't add a `Distinct`) in the loops.
I thought the risk low enough but now realized that a control can register and a control can be used many times..

I'll update it to ConcurrentDictionary to guarantee uniqueness. 



---
In reply to: [420391613](https://github.com/microsoft/react-native-windows/pull/4793#discussion_r420391613) [](ancestors = 420391613)

Reply:
Now the only race is that we could throw an error when an assemlby is actually added, but that implies the code was not properly synchronized so it is okay to fail anyway.

---
In reply to: [420415841](https://github.com/microsoft/react-native-windows/pull/4793#discussion_r420415841) [](ancestors = 420415841,420391613)